### DISCUSSION TITLE
make useMachine hook available for other hook compatible libraries

### DISCUSSION THREAD

**Discussion by herrKlein** [statelyai/xstate#3427] [Ideas]
We share our 'react' hooks between 'preact' and 'haunted' or even 'solid'. Those libraries have support for react Hooks. We ourselves love preact and haunted because they are more compatible with integrating webcomponents then react.

It would be supportive that the useMachine hook, can also work within those libraries.

for now it's not easy mainly because of proprietary react hooks like: 'useSyncExternalStoreWithSelector'.

Is it possible ( and even simply by creating a library agnostic useMachine hook ) to support those libraries?

thanks in advance!

**Comment by Andarist** [comment]
> and even simply by creating a library agnostic useMachine hook

I don't think this is a good strategy here. For an improved integration with React, we need to use `useSyncExternalStoreWithSelector` and that won't be removed from the internals. So since `@xstate/react` has to stay React-specific there is no real gain here to be had from an "agnostic" library. Each framework has some unique APIs and behaviors, only relevant to it. So from our perspective, it's best to provide a specialized package for each framework.

**Reply by herrKlein** [reply]
I'll understand
How difficult would it be to port the @xstate/react ( of version 17, without the useSyncExternalStore hook )
package to a hook compatible library? Do you know of any other react specific hooks used?

First choice for us would be haunted because it has support for web-components _and_ virtual components.

**Reply by Andarist** [reply]
It shouldn't be overly complicated to port `@xstate/react` to another library. In fact, we weren't using `useSyncExternalStore` up until React 18 has been released so the older implementation can still be found in older versions of @xstate/react`.
